gdx/src/com/badlogic/gdx/math/Octree.java

```diff
@@ -0,0 +1,382 @@
+ *       return tmp.dst2(ray.origin);
+ *     }
+ *     return Float.MAX_VALUE;
+ *   }
+ * });
+ *
+ * // Adding game objects to the octree
+ * octree.add(gameObject1);
+ * octree.add(gameObject2);
+ *
+ * // Querying the result
+ * ObjectSet<GameObject> result = new ObjectSet<>();
+ * octree.query(cam.frustum, result);
+ *
+ * // Rendering the result
+ * for (GameObject gameObject : result) {
+ *   modelBatch.render(gameObject);
+ * }
+ * </pre>
+ */
+public class Octree<T> {
+
+	private final int maxItemsPerNode;
+
+	private final Pool<OctreeNode> nodePool = new Pool<OctreeNode>() {
+		@Override
+		protected OctreeNode newObject() {
+			return new OctreeNode();
+		}
+	};
+
+	protected OctreeNode root;
+	private final Collider<T> collider;
+
+	private static final Vector3 tmp = new Vector3();
+
+	public Octree (Vector3 minimum, Vector3 maximum, int maxDepth, int maxItemsPerNode, Collider<T> collider) {
+		super();
+		Vector3 realMin = new Vector3(Math.min(minimum.x, maximum.x), Math.min(minimum.y, maximum.y), Math.min(minimum.z, maximum.z));
+		Vector3 realMax = new Vector3(Math.max(minimum.x, maximum.x), Math.max(minimum.y, maximum.y), Math.max(minimum.z, maximum.z));
+
+		this.root = nodePool.obtain();
+		this.root.bounds.set(realMin, realMax);
+		this.root.level = maxDepth;
+		this.collider = collider;
+		this.maxItemsPerNode = maxItemsPerNode;
+	}
+
+	public void add (T object) {
+		root.add(object);
+	}
+
+	public void remove (T object) {
+		root.remove(object);
+	}
+
+	public void update (T object) {
+		root.remove(object);
+		root.add(object);
+	}
+
+	/** Method to retrieve all the geometries.
+	 * @param resultSet
+	 * @return the result set
+	 */
+	public ObjectSet<T> getAll (ObjectSet<T> resultSet) {
+		root.getAll(resultSet);
+		return resultSet;
+	}
+
+	/** Method to query geometries inside nodes that the aabb intersects.
+	 * Can be used as broad phase.
+	 * @param aabb - The bounding box to query
+	 * @param result - Set to be populated with objects inside the BoundingBoxes
+	 */
+	public ObjectSet<T> query (BoundingBox aabb, ObjectSet<T> result) {
+		root.query(aabb, result);
+		return result;
+	}
+
+	/** Method to query geometries inside nodes that the frustum intersects.
+	 * Can be used as broad phase.
+	 * @param frustum - The frustum to query
+	 * @param result set populated with objects near from the frustum
+	 */
+	public ObjectSet<T> query (Frustum frustum, ObjectSet<T> result) {
+		root.query(frustum, result);
+		return result;
+	}
+
+	public T rayCast (Ray ray, RayCastResult<T> result) {
+		result.distance = result.maxDistanceSq;
+		root.rayCast(ray, result);
+		return result.geometry;
+	}
+
+	/** Method to get nodes as bounding boxes. Useful for debug purpose.
+	 *
+	 * @param boxes
+	 */
+	public ObjectSet<BoundingBox> getNodesBoxes (ObjectSet<BoundingBox> boxes) {
+		root.getBoundingBox(boxes);
+		return boxes;
+	}
+
+	protected class OctreeNode {
```

Comment:
Is there a reason why is this nested class not static? If it was, the node pool could be static as well and I suspect that it would allow `children` to be raw array as well.

Reply:
we need to avoid garbage as much as possible especially for dynamic stuff, children as raw array wouldn't be better.

sharing pool across multiple Octrees could be an issue, geometry array could be small in some case an big in other case which could lead to unnecessary heavy allocations. This could happen within a single octree though but it would be less annoying (isolated) than with a shared pool.

Reply:
> we need to avoid garbage as much as possible especially for dynamic stuff, children as raw array wouldn't be better.

I was referring to the related discussion around the `children` above, I am not arguing for or against that change, as I have not read the code that thoroughly. That comment was only about the static-ness of `OctreeNode`.

Reply:
Sorry i have deep knowledge of that class now :D, so i got directly to the point of shared pool. 
To answer your question, OctreeNode use some information from Octree (collider, maxItems), we could pass them to OctreeNode, but not sure it worth it.